WEBDOM-390: Provide a basic framework to support clearing cache during provisioning.

Jelle-S opened the pull request.

Jelle-S committed:
WEBDOM-390: Provide a basic framework to support clearing cache during provisioning.

Jelle-S committed:
WEBDOM-390: Code style fixes.

Jelle-S committed:
Fix PHPUnit version constraint.

Jelle-S committed:
WEBDOM-390: Fix tests.

dubgeiser commented on CLI/DefaultCliFactory.php:
```diff
@@ -0,0 +1,53 @@
+            $key->loadKey(file_get_contents($keyLocation));
+
+            if (!$ssh->login($user, $key)) {
+                throw new \Exception('SSH login failed.');
```
Is there a way we could provide more information?  (At least mention which $user?)

dubgeiser commented on CLI/RemoteCli.php:
```diff
@@ -0,0 +1,28 @@
+    {
+        $this->connection = $connection;
+        if ($cwd) {
+            $this->connection->exec('cd -P ' . escapeshellarg($cwd));
```
```php
$this->execute('cd -P ' . escapeshellarg($cwd));
```
seems more OOPie

dubgeiser commented on CLI/CliInterface.php:
```diff
@@ -0,0 +1,8 @@
+
+interface CliInterface
+{
+    public function execute(string $command);
```
Maybe add a comment that $command should be properly shell-escaped?

dubgeiser commented on CacheClearer/CacheClearerInterface.php:
```diff
@@ -0,0 +1,10 @@
+
+interface CacheClearerInterface
+{
+    public function clearCache($object, CliInterface $cli);
```
Missing type declaration of $object?

Jelle-S replied:
$object could theoretically be anything, cache can be cleared for an application environment, for en entire environment, for only one server in an environment, ...

dubgeiser commented on Provider/CacheClearProvider.php:
```diff
@@ -0,0 +1,34 @@
+        $class = get_class($object);
+
+        if (!$class || !isset($this->cacheClearers[$class])) {
+            throw new NoCacheClearerFoundException('No cache clearer found for ' . $class);
```
suggestion: introduce finer grained exception handling so it is more clear that either $object is null or not an instance and a separate exception for not finding a cache clearer.

dubgeiser commented on Provider/CliFactoryProvider.php:
```diff
@@ -0,0 +1,51 @@
+
+        if (!$class || !isset($this->cliFactories[$class])) {
+            if (!($this->defaultCliFactory instanceof CliFactoryInterface)) {
+                throw new NoCliFactoryFoundException('No cli factory found for ' . $class);
```
More meaningful error message might be desired.

dubgeiser commented on Provisioner/CacheClearBuildProvisioner.php:
```diff
@@ -0,0 +1,54 @@
+            // There is no cli factory registered for this application
+            // envrironment, meaning we can't clear cache for it. This probably
+            // shouldn't make the task fail, but should we log it somehow?
+        }
```
Since I'm missing (a lot of) context, I have no idea about logging stuff here, but smothering exceptions never seems like a good idea.

dubgeiser reviewed (commented):
Just commenting.  I have not reviewed the test cases.

Jelle-S committed:
WEBDOM-390: Refactoring based on review comments.

- Written better exception message when SSH login fails.
- Used the execute method in the RemoteCli constructor.
- Added docblock for CliInterface::execute().
- Introduced finer grained exception handling in CacheClearProvider.
- Introduced finer grained exception handling in CliFactoryProvider.
- Added logging to CacheClearBuildProvisioner.

Jelle-S committed:
WEBDOM-390: Improved error handling and logging.

dubgeiser reviewed (approved):
Thank you!

Jelle-S committed:
WEBDOM-390: Code style fix.

daften commented on CLI/DefaultCliFactory.php:
```diff
@@ -0,0 +1,55 @@
+
+class DefaultCliFactory implements CliFactoryInterface
+{
+    public function create($object): ?CliInterface
```
Docblock

daften commented on CLI/DefaultCliFactory.php:
```diff
@@ -0,0 +1,55 @@
+                . ', ' . get_class($object) . ' given.'
+            );
+        }
+        $appEnv = $object;
```
Why call it object and don't typehint, if it's an app environment?

matthijs-va replied:
Probably because he plans on supporting other objects as well.

Jelle-S replied:
Because we have to implement the interface, and you could theoretically create a cli for an Environment (if you have a root user and need to do something for the entire env while provisioning), VirtualServer, ... . So it can be anything, and Doctrine/Symfony entities don't have a common class they inherit from, so we can't typehint that either.

daften commented on CLI/RemoteCli.php:
```diff
@@ -0,0 +1,64 @@
+     */
+    protected $lastOutput;
+
+
```
unnecessary whiteline?

daften reviewed (changes requested):
Waar zijn de tagged services met tag domainator.cacheclearer bv.

Jelle-S commented:
@daften RE services tagged with domainator.cacheclearer: They are apptype-dependent, for example `bin/console cache:clear --env=prod` for symfony, `drush cr all` for D8, `drush cc all` for D7, ... So they are defined in those bundles. @matthijs-va suggested adding an opcode cache clearer in core, but not everyone has https://github.com/gordalina/cachetool installed so I'm not sure if that's the best idea... unless we have some way to opt-out...

Jelle-S committed:
WEBDOM-390: Added docblock per @daften's request.

Jelle-S committed:
WEBDOM-390: Fixed review remarks.

- Added docblocks.
- Removed unnecessary whiteline.

Jelle-S committed:
WEBDOM-390: Use the command builder to execute commands.

Jelle-S committed:
WEBDOM-390: Fix tests.

Jelle-S committed:
Log cli output when clearing cache.
